librubyfmt/src/format.rs

```diff
@@ -2807,287 +2744,155 @@ fn can_elide_parens_for_reserved_names(cc: &[CallChainElement]) -> bool {
+        let is_last_call_args = if let Some(last_call_index) = last_call_index {
+            index == (last_call_index + 1)
+        } else {
+            false
+        };
 
         match cc_elem {
             CallChainElement::Paren(p) => format_paren(ps, p),
             CallChainElement::IdentOrOpOrKeywordOrConst(i) => {
                 let ident = i.into_ident();
-                element_is_super_keyword = ident.1 == "super";
+                element_must_use_parens = ident.1 == "super";
 
-                format_ident(ps, ident)
+                if ident.1 == ".()" {
+                    ps.emit_ident(".".to_string());
+                    element_must_use_parens = true;
+                } else {
+                    format_ident(ps, ident);
+                }
+                ps.shift_comments();
             }
             CallChainElement::Block(b) => {
                 ps.emit_space();
                 format_block(ps, b)
             }
-            CallChainElement::VarRef(vr) => format_var_ref(ps, vr),
+            CallChainElement::VarRef(vr) => {
+                format_var_ref(ps, vr);
+                ps.shift_comments();
+            }
             CallChainElement::ArgsAddStarOrExpressionListOrArgsForward(aas, start_end) => {
+                let end_line = start_end.map(|se| se.1);
+
                 if !aas.is_empty() || next_args_list_must_use_parens {
-                    let delims = if elide_parens {
-                        BreakableDelims::for_kw()
+                    let use_parens = if next_args_list_must_use_parens {
+                        true
+                    } else if is_last_call_args && last_call_use_parens.is_some() {
+                        last_call_use_parens.unwrap()
                     } else {
+                        !elide_parens
+                    };
+                    let delims = if use_parens {
                         BreakableDelims::for_method_call()
+                    } else {
+                        BreakableDelims::for_kw()
                     };
 
-                    let end_line = start_end.map(|se| se.1);
+                    // For def visiblity modifiers, e.g. `private def ...`
+                    if args_has_single_def_expression(&aas) {
+                        ps.emit_space();
 
-                    ps.breakable_of(
-                        delims,
-                        Box::new(|ps| {
-                            format_list_like_thing(ps, aas, end_line, false);
-                        }),
-                    );
-                    if let Some(end_line) = end_line {
-                        // If we're rendering a single-line chain, force a reset so
-                        // that comments end up at the current indentation level
-                        if !render_multiline_chain {
+                        if let ArgsAddStarOrExpressionListOrArgsForward::ExpressionList(mut el) =
+                            aas.clone()
```

Comment:
is the clone call here to prevent the `pop` from modifying? If so can we add a comment?